src/linter.cc

```diff
@@ -22,92 +22,245 @@
+                    sql[i] != '\n' ) {
+                ++i;
+            }
+        }
 
-    absl::Status parser_status = zetasql::ParseStatement(sql,
-            zetasql::ParserOptions(), &output);
+        if (!inside_string && sql[i-1] == '/' && sql[i] == '/') {
+            includes_type2 = true;
+            // ignore the line.
+            while ( i < static_cast<int>(sql.size()) &&
+                    sql[i] != '\n' ) {
+                ++i;
+            }
+        }
 
-    if ( !parser_status.ok() )
-        return parser_status;
-    // TODO(orhan_uysal): Implement KeywordExtractor.
+        if (!inside_string && sql[i] == '#') {
+            includes_type3 = true;
+            // ignore the line.
+            while ( i < static_cast<int>(sql.size()) &&
+                    sql[i] != '\n' ) {
+                ++i;
+            }
+        }
 
-    for (const zetasql::ASTNode *keyword : keywords) {
-        if ( !allUpperCase(keyword) ) {
-            return absl::Status(
-                absl::StatusCode::kFailedPrecondition, "");
+        // ignore multiline comments.
+        if (!inside_string && sql[i-1] == '/' && sql[i] == '*') {
+            // it will start checking after '/*' and after the iteration
+            // finished, the pointer 'i' will be just after '*/' (incrementation
+            // from the for statement is included).
+            i += 2;
+            while ( i < static_cast<int>(sql.size()) &&
+                    !(sql[i-1] == '*' && sql[i] == '/') ) {
+                ++i;
+            }
         }
+
+        if (sql[i] == '\'' || sql[i] == '"')
+            inside_string = !inside_string;
     }
 
+    if ( includes_type1 + includes_type2 + includes_type3 > 1 )
+        return absl::Status(
+                absl::StatusCode::kFailedPrecondition,
+                "either '//' or '--' should be used to "
+                "specify a comment");
     return absl::OkStatus();
 }
 
-}  // namespace linter
-}  // namespace zetasql
+absl::Status ASTNodeRule::ApplyTo(absl::string_view sql) {
+    RuleVisitor visitor(rule_, sql);
+
+    std::unique_ptr<ParserOutput> output;
+    ParseResumeLocation location =
+        ParseResumeLocation::FromStringView(sql);
+
+    bool is_the_end = false;
+    while (!is_the_end) {
+        ZETASQL_RETURN_IF_ERROR(ParseNextScriptStatement(
+            &location, ParserOptions(), &output, &is_the_end));
+        ZETASQL_RETURN_IF_ERROR(
+            output->statement()->TraverseNonRecursive(&visitor));
+    }
+
+    ZETASQL_RETURN_IF_ERROR(visitor.GetResult());
+
+    return absl::OkStatus();
+}
+
+absl::Status CheckAliasKeyword(absl::string_view sql) {
+    return ASTNodeRule([](const ASTNode* node,
+        absl::string_view sql) -> absl::Status {
+        if (node->node_kind() == AST_ALIAS) {
+            int position = node->GetParseLocationRange()
+                .start().GetByteOffset();
+            if ( sql[position] != 'A' || sql[position+1] != 'S' ) {
```

Comment:
If alias is a single character, this will crash (will it?). Try with:
```
SELECT 1 a;
```

Reply:
I think there is always at least '\0' at the end of the string and at least one letter for alias, so its fine.